NUX: Make tips collapsed by default

noisysocks opened the pull request:
Closes #9242. Fixes #12358.

![nux](https://user-images.githubusercontent.com/612155/48600251-44f91500-e9bf-11e8-90fe-9b417a595904.gif)

Makes the NUX guide less overwhelming by collapsing the first tip in the guide.

What's changed:

- `DotTip` has a new `isCollapsible` prop which marks it as being a collapsible tip
  - Collapsible tips are opened and closed by clicking on the indicator
  - I've increased the size of the indicator so that it's easier to click
  - The first tip is marked as `isCollapsible`
- Dismissing and disabling tips works as before

#### How to test

1. Clear out your local storage or re-enable tips
2. Note that the first tip is now a collapsed pulsating dot
3. Click on it to open up the first tip
4. Click through the rest of the tips

noisysocks committed:
NUX: Make tips collapsible

- Adds `isCollapsible` to `DotTip` so as to support collapsible tips.
  These are tips that only show the indicator dot. When the indicator dot
  is clicked, the tip opens like normal.
- Make the NUX guide less overwhelming by collapsing the first tip in
  the guide.

karmatosed commented:
Thanks for this. I do agree that having the pulsing is much better. Unfortunately in testing though there are a few issue that to me mean we have to iterate:

![1](https://user-images.githubusercontent.com/253067/47873358-ecdfe180-de08-11e8-90c6-11e277c98156.png)

Ideally we would I think only have in a walkthrough as the Tips are today, one showing at a time. This multiple pulsing being the first experience is too much.

![2](https://user-images.githubusercontent.com/253067/47873399-041ecf00-de09-11e8-97fa-0658d87560ca.png)

Whilst this goes when we only have one by one, having one Tip open is a weird experience right now. For now having it just become one by one works. Pulsing one opens to show Tip and then when you close a new one pulses.

![3](https://user-images.githubusercontent.com/253067/47873457-27e21500-de09-11e8-9d54-bb847c332b92.png)

As illustrated above, having multiple Tips open is easy with them all showing. This isn't a great experience so let's step back to just 1 by 1 in a flow.

That all said, I do think the pulsing itself is much better. My issue is with the number :)

karmatosed reviewed (changes requested):
As per my comment, let's have only one at a time for now.

noisysocks commented:
@karmatosed: I've made it so that only the first tip is collapsed. Give me a 🙂 or a 🙁 and I'll work on making the tests pass so that we can merge this.

karmatosed commented:
Thanks, that's a lot better!

noisysocks commented:
Thanks @karmatosed! ❤️ 

I've made the tests pass, updated the PR description, and tagged some folks for a code review.

tofumatt reviewed (commented):
This seems like quite an aggressive change to be making well after a UI freeze and so close to the 5.0 release.

There are probably accessibility concerns with closing these by default. What's the intended UX for screenreaders, for instance? Do we want the tips to need to be "activated" by screenreader users (I don't think that's ideal/good UX) or should they be active by default in the DOM/screenreader UX and just visually hidden (I think that's better)?

I'll need to do a fair amount of testing on this, later in the day.

karmatosed commented:
@tofumatt having multiple tips and having them open by default also has accessible issues. The goal here is to make it more usable. What can be done with it closed to pass a11y?

tofumatt commented:
If there will be multiple tips present in the UI with only a visual indicator of its relation to the UI, a screenreader won't be able to contextualise it. So each tip should *have a label/short descriptor/intro to what the tip is about* so a screenreader user can choose to expand it. That would be a similar UX for a screenreader and allow the tip to be hidden by default (and also allow a few on-screen at the same time without confusion).

If we can't do that timing-wise... it should probably be the case that screenreader users don't have to "Open" the tips as that's a visual UX—the tip is visually hidden to reduce visual clutter, but for a screenreader it can't relate the tip to the visual UI. Best to just read it out.

Does that make sense?

tofumatt reviewed (changes requested):
Maybe I'm missing something, but how are these tooltips announced to users of assistive technology? I don't see them being announced and I can't see how to access them with the keyboard when they're hidden. Additionally: when I clicked on one the screenreader didn't read out the contents.

These seem like an accessibility regression, which is a real shame because these tips would be useful for explaining things like keyboard shortcuts to users.

When I open a page with tips on `master`, the tips are read out once I tab at least... it's not great but it works. On this branch I can't tab to the tips obviously and they're basically lost on a keyboard/screenreader user.

I would think that the best way forward would be to have tips use `wp.a11y.speak( $short_tip_description )` to announce themselves to screenreader users and provide an obvious way to access them with a keyboard. Maybe include a keyboard shortcut to open them in the short description.

karmatosed commented:
@tofumatt can they announce and be closed? Just want to work out boundaries here.

tofumatt commented:
Right now when there are “closed”, pulsing tips on screen there is no indication to a screenreader user.

If tips are on-screen, even closed, we should tell the user:

1. There are tips
2. How to open them
3. How to dismiss them (optional but nice)

So they can start closed, but we need a way to announce their presence if not their contents, and a way for keyboard users to access them and open them.

mtias commented:
This seems like it needs a bit more work to cover all the cases.

noisysocks commented:
Picking this back up. It's something that I think we should do as part of 5.0.1, which means that there's some time to iterate on a good solution here.

The button which opens a collapsed tip appears in the DOM like any other button. When I enable VoiceOver, check this branch out, and then tab through the toolbar, I can hear VoiceOver read:

- "Add block, selected, button"
- "Open tip, selected, button"
- "Undo, selected, button"

Would changing the label of the _Open tip_ button to something more explanatory suffice? e.g. `"Open tip for 'Add block'"`

I get your point about it not being immediately obvious to users of screen readers that there are tips visible on the screen. I'll need your guidance here as I'm not familiar with `wp.a11y.speak()` best practices. Would `speak()`ing something like "There is one tip on the screen. Press ^⌥t to change focus to the next tip" when the page loads work?

Stepping back, I'm not clear on how best to make the NUX guide useful to users of screen readers since much of it is premised on visually pointing to UI elements. I could use your guidance there, too.

karmatosed commented:
I agree @noisysocks. I do think it's not great they are how they are right now but it is something that can be iterated.

tofumatt commented:
> "Open tip for 'Add block'"

works for me. 👍

Announcing there are tips available on page load might not be great actually after thinking on it because `speak()` is largely used to convey information changed after interaction that a screenreader won't otherwise pick up on.

noisysocks committed:
Use more descriptive label for the 'Add block' collapsible tip

noisysocks commented:
I've updated this to use a more descriptive label for screen readers.

noisysocks commented:
I think this is ready to go in. Just needs a code review. No rush though since it's targeted for 5.0.1 🙂

tofumatt reviewed (changes requested):
The issue with this is there still isn't an announcement of the fact that there *are* tips anymore with them hidden. There's no awareness that a tip exists (or how to activate it) for a screenreader user.

In `master` the tip, in its entirely, is read aloud to the user. Here's NVDA's speech log with it in action:

<img width="1356" alt="screenshot 2018-12-03 20 20 47" src="https://user-images.githubusercontent.com/90871/49399349-f5457800-f738-11e8-9afd-9745653bdd2a.png">

In this branch there's nothing like it so there's just no context for a tip existing.

What we need is for there to be HTML in the DOM that is read out by a screenreader instructing the user that there is a tip available and how to access it. The old tip's content were read out by the screenreader... if that didn't work then wherever the code that checks if tips should be displayed _should_ probably use `speak()` to say something like `speak( 'Tips on new features are available; press $KEYBOARD_SHORTCUT to open them.' )` or something like that.

afercia commented:
I see only the first tip is a button. All the other ones are divs, so their popover can't be closed and have the pulsing tip still active. At that point, users have either to go through all the tips or dismiss them permanently.

Basically this proposed change collapses only the first tip. Not sure what the added value is other than saving a few clicks. Seems to me this change tries to solve a discoverability problem by means of a more complicated and less intuitive interaction with the tips.

That said, re: the first tip (the button one): while it has an `aria-label`, it also needs to expose its name. This was mentioned several times on other issues. For example, how a speech input users is supposed to activate the first button? The accessible name (the aria-label) of the button is not exposed visually in any way:

<img width="354" alt="screenshot 2018-12-04 at 18 05 18" src="https://user-images.githubusercontent.com/1682452/49460216-1cf81700-f7f1-11e8-8199-b0d831ca6f24.png">

For normal buttons, at least we have tooltips (which are already a big trade-off):

<img width="360" alt="screenshot 2018-12-04 at 18 05 29" src="https://user-images.githubusercontent.com/1682452/49460236-284b4280-f7f1-11e8-90c1-dfbc3bb38171.png">

Pre-existing issues:
- the animation on the pulsing tip makes the CPU fan go crazy after a couple minutes, even on a MacBook Pro. Wondering what happens on less performant hardware. As this change allows users to keep the first tip active and pulsing indefinitely in the UI, it's very likely we're going to stress their CPU
- in some steps, the tooltip of the "X" close button is still hidden behind the popover:

<img width="574" alt="screenshot 2018-12-04 at 18 21 19" src="https://user-images.githubusercontent.com/1682452/49460871-abb96380-f7f2-11e8-940e-1fd0db79e79d.png">

noisysocks committed:
Liberate the label copy

noisysocks committed:
NUX: Allow collapsed tips to be opened with the keyboard

Attempts to improve the accessibility of tips by:

- Allowing `DotTip` to be expanded or collapsed using a keyboard
  shortcut configured with the `shortcut` prop.
- Having collapsed tips that have a shortcut configured `speak()` when
  mounted so that screen reader users are aware of their presense.

noisysocks commented:
Thanks for the feedback everyone!

> What we need is for there to be HTML in the DOM that is read out by a screenreader instructing the user that there is a tip available and how to access it.

I've had a first pass at doing this in 5fa8316757fd73fef1e053a7ee9ce515ba4d556c. Let me know what you think and we can iterate on the approach 🙂

> Basically this proposed change collapses only the first tip. Not sure what the added value is other than saving a few clicks.

cc. @karmatosed who can explain the motivation here. My understanding is that we want to make a user's first impression of the new editor less overwhelming by not having the first tip appear so intrusively on the screen. 

> That said, re: the first tip (the button one): while it has an aria-label, it also needs to expose its name.

Good catch—will fix this up tomorrow.

> Pre-existing issues

Are there issues for these? They sound like bugs that we should fix regardless of whether this PR lands.

noisysocks committed:
NUX: Add tooltip to button which collapses and expands collapsible tips

afercia commented:
> Are there issues for these? They sound like bugs that we should fix regardless of whether this PR lands.

- first one (tip animation / CPU load): I don't think there's an issue for that. Maybe it was mentioned in another issue but I'd like to be confirmed before creating an issue,. Would be great to test on a medium / cheap hardware 
- second one (tooltips hidden behind popover): https://github.com/WordPress/gutenberg/issues/7562 created on June

> make a user's first impression of the new editor less overwhelming by not having the first tip appear so intrusively on the screen.

I understand the goal of this PR 🙂 I'm just saying that it makes things more complicated, interaction becomes unexpected, at the point I wonder if tips still honor the reason they were implemented for. The original goal was: [on first load you see a message that encourages you to discover more about the new experience](https://github.com/WordPress/gutenberg/issues/3670#issue-277004909). Now instead we want to hide them ¯\_(ツ)_/¯

noisysocks commented:
After speaking with @karmatosed, we're shelving this PR for now. In the future we would like to revisit the UX of tips more whole holistically.
